rootfs/etc/nginx/lua/balancer/sticky.lua

```diff
@@ -151,30 +130,34 @@ function _M.balance(self)
 
   local new_upstream
 
-  new_upstream, key = pick_new_upstream(self)
+  new_upstream, key = self:pick_new_upstream(get_failed_upstreams())
   if not new_upstream then
     ngx.log(ngx.WARN, string.format("failed to get new upstream; using upstream %s", new_upstream))
   elseif should_set_cookie(self) then
-    set_cookie(self, key)
+    self:set_routing_key(key)
   end
 
   return new_upstream
 end
 
 function _M.sync(self, backend)
+  -- reload balancer nodes
   balancer_resty.sync(self, backend)
 
   -- Reload the balancer if any of the annotations have changed.
   local changed = not util.deep_compare(
     self.cookie_session_affinity,
     backend.sessionAffinityConfig.cookieSessionAffinity
   )
+
   if not changed then
     return
   end
 
   ngx_log(INFO, string_format("[%s] nodes have changed for backend %s", self.name, backend.name))
 
+  self.traffic_shaping_policy = backend.trafficShapingPolicy
+  self.alternative_backends = backend.alternativeBackends
```

Comment:
Moving these to `sync` functions seems to be OK because you call this `sync` function from `new` functions in sticky balancer implementations. However I think this introduces a bug:

Imagine a scenario where list of endpoints changes, alternatives backends get updated but affinity annotations stay unchanged. In this case we will not update `self.alternative_backends` because we will short-circuit in https://github.com/kubernetes/ingress-nginx/pull/4514/files#diff-8f83606fdd5dc8d2406cda71b9609762R154. I think we should move these two lines to above https://github.com/kubernetes/ingress-nginx/pull/4514/files#diff-8f83606fdd5dc8d2406cda71b9609762R148

Reply:
> Imagine a scenario where list of endpoints changes, alternatives backends get updated but affinity annotations stay unchanged.

@ElvinEfendi 
Actually, is there any good reason to keep the deep_compare at all? I would imagine that this is much more expensive than just updating self.cookie_session_affinity. After all we don't reinitialize anything if that configuration changes, we just assign a variable if it changed.

I will just remove the deep_compare. If you are not happy with that, I can always change it back later.